lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -1288,11 +1316,57 @@ instance FromJSON WalletPutPassphraseData where
 instance ToJSON  WalletPutPassphraseData where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+instance FromJSON (ApiT PoolMetadataGCStatus) where
+    parseJSON = withObject "PoolMetadataGCStatus" $ \o -> do
+        (status' :: String) <- o .: "status"
+        last_run <- o .:? "last_run"
+        case (status', last_run) of
+            ("restarting", Just (ApiT (Iso8601Time gctime)))
+                -> pure $ ApiT (Restarting $ utcTimeToPOSIXSeconds gctime)
+            ("has_run", Just (ApiT (Iso8601Time gctime)))
+                -> pure $ ApiT (HasRun $ utcTimeToPOSIXSeconds gctime)
+            ("restarting", Nothing)
+                -> fail "missing field last_run"
+            ("has_run", Nothing)
+                -> fail "missing field last_run"
+            ("not_applicable", _)
+                -> pure $ ApiT NotApplicable
+            ("not_started", _)
+                -> pure $ ApiT NotStarted
+            _ -> fail ("Unknown status: " <> status')
+
+instance ToJSON (ApiT PoolMetadataGCStatus) where
+    toJSON (ApiT (NotApplicable)) =
+        object [ "status" .= String "not_applicable" ]
+    toJSON (ApiT (NotStarted)) =
+        object [ "status" .= String "not_started" ]
+    toJSON (ApiT (Restarting gctime)) =
+        object [ "status" .= String "restarting"
+            , "last_run" .= ApiT (Iso8601Time (posixSecondsToUTCTime gctime)) ]
+    toJSON (ApiT (HasRun gctime)) =
+        object [ "status" .= String "has_run"
+            , "last_run" .= ApiT (Iso8601Time (posixSecondsToUTCTime gctime)) ]
```

Comment:
**Suggestion:** Rather than defining these instances by hand, consider making an API type that matches the shape of your on-the-wire format, and then using the generic deriving mechanism.

```hs
data ApiPoolMetadataGCStatus = ApiPoolMetadataGCStatus
    { status :: ApiPoolMetadataGCStatusOption
    , lastRun :: Maybe (ApiT POSIXTime)
    } deriving (Eq, Generic, Show)

data ApiPoolMetadataGCStatusOption
    = NotApplicable
    | NotStarted
    | Restarting
    | HasRun
    deriving (Eq, Show, Generic)
```

Then you can write:

```hs
instance FromJSON ApiPoolMetadataGCStatus where
    parseJSON = genericParseJSON defaultRecordTypeOptions
instance ToJSON ApiPoolMetadataGCStatus where
    toJSON = genericToJSON defaultRecordTypeOptions

instance FromJSON ApiPoolMetadataGCStatusOption where
    parseJSON = genericParseJSON defaultRecordTypeOptions
instance ToJSON ApiPoolMetadataGCStatusOption where
    toJSON = genericToJSON defaultRecordTypeOptions

instance FromJSON (ApiT POSIXTime) where
    parseJSON = ...
instance ToJSON (ApiT POSIXTime) where
    toJSON = ...
```

And then you'd just need to provide a wrapper function:
```hs
mkApiPoolMetadataGCStatus :: PoolMetadataGCStatus -> ApiPoolMetadataGCStatus
```

We actually use this pattern fairly commonly throughout the codebase. (Search for `mkApi` to see lots of examples.)

Reply:
I kind of prefer to have types be the way I want them in the code, which is the current representation with the more expressive sum type. JSON is a little less expressive and that is fine. The JSON instances are exactly what is supposed to fill the gap. We have json roundtrip tests to ensure these are all well-formed. Writing manual instances is fine, IMO.

Reply:
> I kind of prefer to have types be the way I want them in the code, which is the current representation with the more expressive sum type. 

I wasn't suggesting to remove your existing `PoolMetadataGCStatus` type. Only to provide an additional API-specific type that matches the shape of your on-the-wire format, and to provide a function from your expressive sum type to the API type:

```hs
mkApiPoolMetadataGCStatus :: PoolMetadataGCStatus -> ApiPoolMetadataGCStatus
```

Reply:
That's basically more code :)

We have boilerplate either way... but if it's in the JSON instance instead of `mkApiPoolMetadataGCStatus`, then we have one type less to maintain.

Reply:
@hasufell I think you're right that it's strictly more _lines_ of code to use generic deriving, but I think it gives us a cleaner design, because:

* We don't have to manually re-implement the bidirectional constructor-name-to-text mappings.
* We don't need to implement JSON serialization for `ApiT POSIXTime` values. Instead, we can just reuse the existing `ApiT Iso8601` instances. 
* It's consistent with what we do for most of the other API types.

Here's a commit which illustrates the point: https://github.com/input-output-hk/cardano-wallet/commit/a57b55db90873ba41cf5109dd01bd1d7fad6a07c.